[bug/fuzzing] "panic: runtime error: nil pointer dereference" when processing ProposerSlashing

# 🐞 Bug Report

### Description

During fuzzing with [beaconfuzz](https://github.com/sigp/beacon-fuzz), I found the following bug:

what: `panic: runtime error: invalid memory address or nil pointer dereference`
where: in `prysm`
how: triggered during `ProposerSlashing` processing.

The bug is happening mainly in `VerifyProposerSlashing` function.


Here is some supposition:
https://github.com/prysmaticlabs/prysm/blob/a0bf8cba523ea08c8d1c199ffce5a7821f789a06/beacon-chain/core/blocks/block_operations.go#L426-L432

- line 426: `proposer` will be `nil` after the call to `ValidatorAtIndexReadOnly`
- line 430: `IsSlashableValidatorUsingTrie` is called with proposer

https://github.com/prysmaticlabs/prysm/blob/9a1157465e9205fbec4026299faf058b8b3bad98/beacon-chain/core/helpers/validators.go#L50-L51

- line 51: `IsSlashableValidatorUsingTrie` will execute `val.Slashed()` (`val` == `proposer`)

https://github.com/prysmaticlabs/prysm/blob/6a9112bfaede08454815a262b733b930f3cf4e22/beacon-chain/state/getters.go#L76-L79

- line 78: `Slashed` try to dereference `v.validator` without having verify that `v` and `v.validator` are not `nil`


## 🔥 Error

Download: [panic_nil_deref_prysm_proposer.zip](https://github.com/prysmaticlabs/prysm/files/4730325/panic_nil_deref_prysm_proposer.zip)


``` sh
./prysm_FuzzProposerSlashing2.libfuzzer panic_nil_deref_prysm_proposer.ssz
INFO: Seed: 1287398888
./prysm_FuzzProposerSlashing2.libfuzzer: Running 1 inputs 1 time(s) each.
Running: panic_nil_deref_prysm_proposer.ssz

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x11d0360]

goroutine 17 [running, locked to thread]:
github.com/prysmaticlabs/prysm/beacon-chain/state.(*ReadOnlyValidator).Slashed(...)
	github.com/prysmaticlabs/prysm/beacon-chain/state/getters.go:78
github.com/prysmaticlabs/prysm/beacon-chain/core/helpers.IsSlashableValidatorUsingTrie(...)
	github.com/prysmaticlabs/prysm/beacon-chain/core/helpers/validators.go:51
github.com/prysmaticlabs/prysm/beacon-chain/core/blocks.VerifyProposerSlashing(0x10c000802f50, 0x10c0000edd50, 0x16, 0x10c000038fe0)
	github.com/prysmaticlabs/prysm/beacon-chain/core/blocks/block_operations.go:430 +0x400
github.com/prysmaticlabs/prysm/beacon-chain/core/blocks.ProcessProposerSlashings(0x18344c0, 0x10c000138030, 0x10c000802f50, 0x10c0000edd80, 0x0, 0x0, 0x0)
	github.com/prysmaticlabs/prysm/beacon-chain/core/blocks/block_operations.go:396 +0x102
github.com/prysmaticlabs/prysm.FuzzProposerSlashing(0x603000001900, 0x20, 0x20, 0x0)
	github.com/prysmaticlabs/prysm/fuzz.go:223 +0x1c5
main.LLVMFuzzerTestOneInput(...)
	command-line-arguments/main.132452089.go:21
main._cgoexpwrap_b2568cfea483_LLVMFuzzerTestOneInput(0x603000001900, 0x20, 0x5ed8ef0e)
	_cgo_gotypes.go:53 +0x56

```

## 🔬 Minimal Reproduction

I can only reproduce the bug running my fuzzer with the previous crashing ssz file.
Maybe you will succeed to reproduce with your fuzzers as well.

Fuzzing function:
``` go
func FuzzProposerSlashing(b []byte) int {
	params.UseMainnetConfig()
	data := &ethpb.ProposerSlashing{}
	if err := data.UnmarshalSSZ(b); err != nil {
		return 0
	}
	// get a valid beaconstate
	st := getbeaconstate()
	s, err := stateTrie.InitializeFromProto(st)
	if err != nil {
		// should never happen
		panic("stateTrie InitializeFromProto")
	}
	// process the container
	post, err := blocks.ProcessProposerSlashings(context.Background(), s, &ethpb.BeaconBlockBody{ProposerSlashings: []*ethpb.ProposerSlashing{data}})
	if err != nil {
		return 0
	}
	if post == nil {
		return 0
	}
	return 1
}
```

I have nevertheless fix the bug (with the following patch) and the fuzzer is not crashing anymore.


## Patch 

A simple fix consist to check if `v` and `v.validator` are `nil` like in `WithdrawableEpoch()` and `ExitEpoch()` functions.

``` go

// Slashed returns the read only validator is slashed.
func (v *ReadOnlyValidator) Slashed() bool {
	if v == nil || v.validator == nil {
		return false
	}
	return v.validator.Slashed
}
```
instead of: https://github.com/prysmaticlabs/prysm/blob/6a9112bfaede08454815a262b733b930f3cf4e22/beacon-chain/state/getters.go#L77-L79

## 🌍  Your Environment

**Operating System:**

OS: Ubuntu 18.04
Go: Go 1.14

**What version of Prysm are you running? (Which release)**

master
commit: d152b4881576b53665ec67aa9968bb8c9d0a1660

--- a/beacon-chain/state/getters.go
+++ b/beacon-chain/state/getters.go
@@ -75,11 +75,17 @@ func (v *ReadOnlyValidator) WithdrawalCredentials() []byte {
 
 // Slashed returns the read only validator is slashed.
 func (v *ReadOnlyValidator) Slashed() bool {
+	if v == nil || v.validator == nil {
+		return false
+	}
 	return v.validator.Slashed
 }
 
 // CopyValidator returns the copy of the read only validator.
 func (v *ReadOnlyValidator) CopyValidator() *ethpb.Validator {
+	if v == nil || v.validator == nil {
+		return nil
+	}
 	return CopyValidator(v.validator)
 }
 

--- a/beacon-chain/state/getters_test.go
+++ b/beacon-chain/state/getters_test.go
@@ -75,3 +75,13 @@ func TestNilState_NoPanic(t *testing.T) {
 	_ = st.FinalizedCheckpoint()
 	_ = err
 }
+
+func TestReadOnlyValidator_NoPanic(t *testing.T) {
+	v := &ReadOnlyValidator{}
+	if v.Slashed() == true {
+		t.Error("Expected not slashed")
+	}
+	if v.CopyValidator() != nil {
+		t.Error("Expected nil result")
+	}
+}

